File: trie/trie3.php
<?php

class Trie {
    private $has_word = false;
    private $children = array();
    public function insert($word, $start=0){
        $tree = $this;
        while($start < strlen($word)){
            if(!isset($tree->children[$word[$start]])){
                $tree->children[$word[$start]] = new Trie();
            }
            $tree = $tree->children[$word[$start]];
            $start++;
        }
        $tree->has_word = true;
    }

    public function search($word, $start=0){
        $tree = $this;
        while($start < strlen($word)){
            if(!isset($tree->children[$word[$start]])){
                return false;
            }else{
                $tree = $tree->children[$word[$start]];
                $start++;
            }
        }
        return $tree->has_word;
    }
}

function test(&$inserts, &$searches){
    $tree = new Trie();
    foreach($inserts as $w){
        $tree->insert($w);
    }
    $count = 0;
    foreach($searches as $w){
        $count += $tree->search($w);
    }
    return $count;
}

$inserts = file($argv[1], FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
$searches = file($argv[2], FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);

echo(test($inserts, $searches));
for($i = 0; $i < 20; $i++){
    test($inserts, $searches);
}
?>

File: trie/trie3.py

class Trie:
    def __init__(self):
        self.has_word = False
        self.children = {}

    def insert(self, word, start=0):
        tree = self
        while start < len(word):
            if word[start] not in tree.children:
                tree.children[word[start]] = Trie()
            tree = tree.children[word[start]]
            start += 1
        tree.has_word = True

    def search(self, word, start=0):
        tree = self
        while start < len(word):
            if word[start] not in tree.children:
                return False
            else:
                tree = tree.children[word[start]]
                start += 1
        return tree.has_word


def test(inserts, searches):
    tree = Trie()
    for w in inserts:
        tree.insert(w)
    count = 0
    for w in searches:
        count += tree.search(w)
    return count


if __name__ == '__main__':
    import sys
    with open(sys.argv[1]) as f:
        inserts = f.read().splitlines()
    with open(sys.argv[2]) as f:
        searches = f.read().splitlines()
    print(test(inserts, searches))
    for _ in range(20):
        test(inserts, searches)
